graph.go

```diff
@@ -0,0 +1,293 @@
+package dalec
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/pkg/errors"
+	"github.com/pmengelbert/stack"
+	"golang.org/x/exp/constraints"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+type Graph struct {
+	target   string
+	specs    map[string]*Spec
+	ordered  orderedDeps
+	indices  map[string]int
+	vertices []*vertex
+	edges    sets.Set[dependency]
+}
+
+type dependency struct {
+	v    *vertex
+	w    *vertex
+	kind string
+}
+
+type cycle []*vertex
+type cycleList []cycle
+type orderedDeps []*Spec
+
+func (o orderedDeps) targetSlice(target ...string) []*Spec {
+	if len(target) == 0 {
+		return []*Spec(o)
+	}
+
+	for i, dep := range o {
+		if dep.Name == target[0] {
+			return o[:i+1]
+		}
+	}
+	return nil
+}
+
+func (g *Graph) Target() *Spec {
+	return g.specs[g.target]
+}
+
+func (g *Graph) Get(name string) (*Spec, bool) {
+	s, ok := g.specs[name]
+	return s, ok
+}
+
+func (g *Graph) OrderedSlice(target ...string) []*Spec {
+	return g.ordered.targetSlice(target...)
+}
+
+func (g *Graph) Len(target ...string) int {
+	if len(target) == 0 {
+		return len(g.ordered)
+	}
+	return len(g.ordered.targetSlice(target[0]))
+}
+
+type vertex struct {
+	name    string
+	index   *int
+	lowlink int
+	onStack bool
+}
+
+var (
+	graphLock  sync.Mutex
+	BuildGraph *Graph
```

Comment:
Personally, I'm a fan of eliminating global state unless it's absolutely necessary, and I don't see a compelling need in this case. It's kind of necessary for registeredHandlers to be global, since the lifetime of that data is the entire run time of a dalec frontend instance. I'm not sure if we can assume that the lifetime of a build graph will be for the entire run of a Dalec frontend?

Reply:
Now that I'm thinking about it, the build graph corresponds to a dalec (multi-)spec, and the frontend will be invoked individually on Dalec spec files, so I guess it's ok lifetime-wise. What I dislike about globals is that it makes the parameterization of functions implicit. I.e., now we have a bunch of handlers that are parameterized on `dalec.BuildGraph` which is data that is built up at runtime, and the behavior will change dramatically based on changes to that global, but the argument doesn't show up in the function signatures or in any of the data structures passed to those functions

Reply:
It will involve a lot of function/method signature changes for several reasons:
- There are no (public) methods on `Graph` that allow mutation of the data structure. Once it has been initialized, it is immutable. I suppose one could mutate the specs returned via pointers, and so we may want to store those as structs rather than pointers.
- Handlers can be generated via a spec (by providing a `frontend` image under a `target`), meaning we have to have the graph at the time the handlers are registered, in order to report accurately on which handlers are available.
- The specs that will be used need to be validated when they are read. In order to know which specs need to be validated, we need to know what specs will actually be used in the build sequence. Another option is to error out even if unused specs have errors.
- The final spec (the spec for the target, or before these changes, the only spec) is de facto global state. Once read and unmarshaled, the spec has the lifespan of the program as it drives all of the actions of the program and needs to be consulted throughout. It's therefore passed around as a pointed from start to finish. The graph technically doesn't need to live past the container generation stage (into the testing stage), but IMO that is splitting hairs.
- A value of `Mutex<T>` can be `const` (and therefore `static`, global, and mutable via mutex) in rust (without unsafe!), because its size and location are known at compile time, access to it is guarded by mutex, and as long as `T` can be initialized at compile time, the initial value is known as well.
- I should note that I agree with you about eliminating global state. I also agree with you on the implicit parameterization issue. I just see other concerns and I don't think the pattern is **that** bad, given the above.

These are reasons. Whether they are compelling or not is up to you :) I'm open to having my mind changed if you see ways around this and a benefit that outweighs the additional work. I also want explicit approval from @cpuguy83 before I change any public function signatures.

Reply:
Graph has been removed from global state